### DISCUSSION TITLE
Is scheduled publish feature in the roadmap now?

### DISCUSSION THREAD

**Discussion by Stupidism** [payloadcms/payload#567] [Feature Requests & Ideas]
What I found is this: https://payloadcms.com/blog/version-0-15-0

<img width="1329" alt="image" src="https://user-images.githubusercontent.com/5801015/168974392-96e5caa2-dc50-4ca4-b1e1-26e6aab7565e.png">

**Comment by DanRibbens** [comment]
This could use a detailed example or made as a plugin, but what you can do is add your own fields with versions with drafts, and manage it through access control to create your own workflows for publishing content.

Say you have a collection for `posts` that includes a field for `publishAt` for scheduling later, you could set it up this way:
```ts
const Post: CollectionConfig = {
  slug: 'posts',
  labels: {
    singular: 'Post',
    plural: 'Posts',
  },
  versions: {
    drafts: true,
  },
  access: {
    read: ({ req: { user } }) => {

      // a logged in user will be able to read all posts
      if (user) {
        return true;
      }

      // a guest will see only posts that have been published for  published at a time before the current time for the request
      return {
        and: [
          {
            publishAt: {
              less_than: new Date().toJSON(),
            },
            _status: {
              equals: 'published',
            },
          },
        ],
      };
    },
    readVersions: ({ req: { user } }) => Boolean(user),
  },
  fields: [
    {
      name: 'publishAt',
      type: 'date',
      admin: {
        position: 'sidebar',
    },
    // ... other fields for the content
  ],
};
```

There are situations that can add complexity to your workflows to plan around:
1. When using a static site generator, posts will only appear after a rebuild so you will need to do some dynamic querying or filtering.
2. If you need to maintain current and future versions of content. For example, you want changes for the home page set to occur at a future date, but your site still needs a published version to be shown before that change and only fetch the newest published version be displayed on the site.

Does that help?

**Reply by DanRibbens** [reply]
You've inspired me to update the public demo to demonstrate the future dated posts example. 👍
https://github.com/payloadcms/public-demo/pull/3
This change will be visible on https://demo.payloadcms.com in the future.

**Reply by Stupidism** [reply]
@DanRibbens I tried this solution, but it doesn't work at all. Are u sure we can query versions in the `access` field? I  think sth is definitely missing here.

**Reply by Stupidism** [reply]
@DanRibbens I created a repro repo here. Can u help me check what's wrong? https://github.com/payloadcms/payload/discussions/702

**Reply by DanRibbens** [reply]
@Stupidism, Thanks for the reproduction steps, that made it easy to see what was going on.

See response: https://github.com/payloadcms/payload/discussions/702#discussioncomment-3063332

**Reply by Stupidism** [reply]
I know I understood it wrongly now. What I want is exactly like you said here

<img width="1063" alt="image" src="https://user-images.githubusercontent.com/5801015/176903384-c7b8eefc-90b6-4f9f-8a4b-3cb0024b915a.png">

Do u have any suggestions on how to achieve this?

What I can imagine is that:

1. Trigger a webhook every time an article is published
2. Add another SchedulePublish Button near the publishDate.   Setup a corn job server to automatically trigger publish when the time comes.

**Reply by DanRibbens** [reply]
> 1. Trigger a webhook every time an article is published

It is not usually recommended to do this. Consider if you have a content change that is being made on multiple pages to or models. In that scenario you need to be able to make a few different edits and then publish the changes all at once. This can be done with a custom UI control for publishing lots of changes in one push. It also cuts down on build if that is a concern from a hosting perspective. If these are non-issue than it can be done this way.

> 2. Add another SchedulePublish Button near the publishDate. Setup a corn job server to automatically trigger publish when the time comes.

You can do this for sure!
Add an `or` condition to your query to accomodate the extra `publishedDate` docs to serve different versions of the document at the correct time. Remember this is separate from the `access` control since we're talking about staticly served pages and that will be different from the user access.

Does that make sense?

**Reply by Stupidism** [reply]
@DanRibbens Yes, I know we need a strategy to group the publish events of articles.

Thanks for your response!

**Comment by Stupidism** [comment]
Yes, it helps.

I’m currently using Contentful and investigating alternatives to that(Just too expensive and hard to customize). 
Our editors like the publish feature they provide very much. It meets both the two situations you mentioned above. 

How to solve them with payloadcms then?

**Reply by Stupidism** [reply]
I think I might need to @DanRibbens to send a notification to you.

**Reply by denolfe** [reply]
Hey @Stupidism , did you take a look at the PR linked in Dan's comment? https://github.com/payloadcms/payload/discussions/567#discussioncomment-2777438

**Reply by Stupidism** [reply]
@denolfe  Yes, it solves the basic requirements of not showing the draft before `publishDate` for the first time.

But for the second situation @DanRibbens mentioned, I think this cannot solve it.

**Reply by Stupidism** [reply]
@denolfe @DanRibbens Thanks, I see the new feature in the demo project now:

<img width="366" alt="image" src="https://user-images.githubusercontent.com/5801015/170957466-5db1207d-1c2e-4a8a-96fd-1622690a5141.png">

**Comment by ssyberg** [comment]
👍🏼 bumping this feature, we've implemented 50+ CMS solutions for clients over the years and this almost always a requirement (fwiw this is a core feature in most CMS solutions I've worked with), would strongly support this to be considered for core and not plugin/roll your own

**Reply by denolfe** [reply]
Tagging this as 'Planned', so we can prioritize accordingly.

**Reply by SalahAdDin** [reply]
Is it now available? @denolfe

**Reply by prineshaz** [reply]
This is extremely important to our customers if we take on Payload.

**Comment by danielmlr** [comment]
any updates on this feature or have i missed the functionality?

**Reply by tigersoft6084** [reply]
@danielmlr I've implemented contentful-like scheduled publish feature, and if you are still in need of it, I can help you.

**Reply by cbratschi** [reply]
How does it work in contentful? For our projects I would need a start and end date i.e. the document is only published in this date range. Both values are optional.

**Reply by wkentdag** [reply]
FYI I built a plug in to do this! https://github.com/wkentdag/payload-plugin-scheduler

Would love some feedback.

**Reply by cbratschi** [reply]
@wkentdag Thanks, looks fine but one has a publish date. An unpublish date is not available.

I need a more generic solution for cron jobs. Some fields have a certain lifetime and Next.js revalidation is needed for such cases. I will have a closer look how to implement this for Payload 2.x and with Payload 3.x running on a custom server.

**Reply by cbratschi** [reply]
Checked the Contentful specs:

https://www.contentful.com/help/scheduled-publishing/

It supports many additional features. Probably too much for most cases and therefore more complicated to use.

**Reply by SalahAdDin** [reply]
It would be nice if Contentful would be open-source :'-)

**Comment by tkrotoff** [comment]
This discussion should probably be closed
Scheduling for publishing has been implemented here: https://github.com/payloadcms/payload/pull/10203

**Reply by SalahAdDin** [reply]
> This discussion should probably be closed Scheduling for publishing has been implemented here: #10203

Going there!

Thanks !
